src/Stack/SourceMap.hs

```diff
@@ -0,0 +1,245 @@
+        Nothing | not alreadyDownloaded && not downloaded -> do
+          logInfo $
+            "Could not find local global hints for " <>
+            RIO.display ac <>
+            ", forcing a redownload"
+          x <- redownload req dest
+          if x
+            then inner True
+            else do
+              logInfo "Redownload didn't happen"
+              pure Nothing
+        _ -> pure mres
+
+    inner2 = liftIO
+           $ Map.lookup ac . fmap (fmap unCabalString . unCabalStringMap)
+         <$> decodeFileThrow (toFilePath dest)
+
+globalsFromDump ::
+       (HasLogFunc env, HasProcessContext env)
+    => ActualCompiler
+    -> RIO env (Map PackageName GlobalPackage)
+globalsFromDump compiler = do
+    let pkgConduit =
+            conduitDumpPackage .|
+            CL.foldMap (\dp -> Map.singleton (dpGhcPkgId dp) dp)
+        toGlobals ds = Map.fromList $ map toGlobal $ Map.elems ds
+        toGlobal d =
+            ( pkgName $ dpPackageIdent d
+            , GlobalPackage (pkgVersion $ dpPackageIdent d))
+    toGlobals <$> ghcPkgDump (whichCompiler compiler) [] pkgConduit
+
+globalsFromHints ::
+       HasConfig env
+    => ActualCompiler
+    -> RIO env (Map PackageName GlobalPackage)
+globalsFromHints compiler = do
+    ghfp <- globalHintsFile
+    mglobalHints <- loadGlobalHints ghfp compiler
+    case mglobalHints of
+        Just hints -> pure $ Map.map GlobalPackage hints
+        Nothing -> do
+            logWarn $ "Unable to load global hints for " <> RIO.display compiler
+            pure mempty
+
+toActual ::
+       (HasConfig env)
+    => SMWanted
+    -> WithDownloadCompiler
+    -> ActualCompiler
+    -> RIO env SMActual
+toActual smw downloadCompiler compiler = do
+    allGlobals <-
+        case downloadCompiler of
+            WithDownloadCompiler -> globalsFromDump compiler
+            SkipDownloadCompiler -> globalsFromHints compiler
+    let globals =
+            allGlobals `Map.difference` smwProject smw `Map.difference` smwDeps smw
+    return
+        SMActual
+        { smaCompiler = compiler
+        , smaProject = smwProject smw
+        , smaDeps = smwDeps smw
+        , smaGlobal = globals
+        }
+
+checkFlagsUsedThrowing ::
+       (MonadIO m, MonadThrow m)
+    => Map PackageName (Map FlagName Bool)
+    -> FlagSource
+    -> Map PackageName ProjectPackage
+    -> Map PackageName DepPackage
+    -> m ()
+checkFlagsUsedThrowing packageFlags source prjPackages deps = do
+    unusedFlags <-
+        forMaybeM (Map.toList packageFlags) $ \(pname, flags) ->
+            checkFlagUsed (pname, flags) source prjPackages deps
+    unless (null unusedFlags) $
+        throwM $ InvalidFlagSpecification $ Set.fromList unusedFlags
+
+checkFlagUsed ::
+       MonadIO m
+    => (PackageName, Map FlagName Bool)
+    -> FlagSource
+    -> Map PackageName ProjectPackage
+    -> Map PackageName DepPackage
+    -> m (Maybe UnusedFlags)
+checkFlagUsed (name, userFlags) source prj deps =
```

Comment:
Just a bit of nitpicking, but for the sake of consistency would it be better to keep it plural: `checkUnusedFlags`?